src/library/scala/collection/mutable/OrderAwareHashMap.scala

```diff
@@ -0,0 +1,800 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.collection
+package mutable
+
+import scala.annotation.meta.{getter, setter}
+import scala.annotation.tailrec
+import scala.collection.generic.DefaultSerializationProxy
+import scala.collection.mutable
+
+/** This class implements mutable sets using a hashtable with red-black trees in the buckets for good
+  * worst-case performance on hash collisions. An `Ordering` is required for the element type. Equality
+  * as determined by the `Ordering` has to be consistent with `equals` and `hashCode`. Universal equality
+  * of numeric types is not supported (similar to `AnyRefMap`).
+  *
+  * @author  Stefan Zeiger
+  * @since   2.13
+  * @see [[http://docs.scala-lang.org/overviews/collections/concrete-mutable-collection-classes.html#hash-tables "Scala's Collection Library overview"]]
+  * section on `Hash Tables` for more information.
+  *
+  * @define Coll `mutable.OrderAwareHashMap`
+  * @define coll mutable order-aware hash set
+  * @define mayNotTerminateInf
+  * @define willNotTerminateInf
+  */
+final class OrderAwareHashMap[K, V](initialCapacity: Int, loadFactor: Double)(implicit ordering: Ordering[K])
+  extends AbstractMap[K, V]
+    with MapOps[K, V, Map, OrderAwareHashMap[K, V]] //--
+    with StrictOptimizedIterableOps[(K, V), Iterable, OrderAwareHashMap[K, V]]
+    with StrictOptimizedMapOps[K, V, Map, OrderAwareHashMap[K, V]] { //--
+
+  def this()(implicit ordering: Ordering[K]) = this(OrderAwareHashMap.defaultInitialCapacity, OrderAwareHashMap.defaultLoadFactor)(ordering)
+
+  import OrderAwareHashMap.Node
+  private[this] type RBNode = OrderAwareHashMap.RBNode[K, V]
+  private[this] type LLNode = OrderAwareHashMap.LLNode[K, V]
+
+  /** The actual hash table. */
+  private[this] var table: Array[Node] = new Array[Node](tableSizeFor(initialCapacity))
+
+  /** The next size value at which to resize (capacity * load factor). */
+  private[this] var threshold: Int = newThreshold(table.length)
+
+  private[this] var contentSize = 0
```

Comment:
I'm not sure this is enough; in the case where there are a lot of elements in very few bins, splitting based on the content size alone isn't likely to be a winning strategy.  Instead, knowing the number of free slots is probably a good thing also (or maybe the only thing to know).  Or maybe we want to know the number of detected hash collisions that can't be resolved by adding more bins.

Reply:
The strategy for expanding the hash table is the same as in the standard `HashMap` where it performs quite well in the benchmarks with the badly distributed hashes. It does mean that hash tables with random bad distributions can become unnecessarily large in both implementations but it's only a linear factor. In a deliberate collision attack you rely on far-worse-than-linear performance degradation from actual hash collisions (not just bucket collisions).